### PR TITLE
Fix quadratic performance with lots of use statements

### LINKED ISSUE #43572: Pathological performance in resolve with lots of unused imports in `style` crate
A [local build](https://pastebin.mozilla.org/9028490) of Servo's `style` crate shows a huge amount of time (46 seconds!) in name resolution. It turns out that [basically all of it](https://pastebin.mozilla.org/9028497) is calling `span_to_snippet`, specifically [this one](https://github.com/rust-lang/rust/blob/2a6828e7f1ed3163a2797f0a111570ef130f7b6b/src/librustc_resolve/check_unused.rs#L150). That's a lot of unused imports!

Sure enough if we create a synthetic file locally with this script:

```bash
N=20000

echo '#![allow(unused)]' > lots-of-unused.rs
for i in `seq 1 $N`; do
  echo "use foo as foo$i;" >> lots-of-unused.rs
done
echo 'fn foo() {}' >> lots-of-unused.rs
echo 'fn main() {}' >> lots-of-unused.rs


echo > lots-of-used.rs
for i in `seq 1 $N`; do
  echo "use test::foo$i;" >> lots-of-used.rs
done
echo 'mod test {' >> lots-of-used.rs
for i in `seq 1 $N`; do
  echo "pub fn foo$i() {}" >> lots-of-used.rs
done
echo '}' >> lots-of-used.rs
echo 'fn main() {' >> lots-of-used.rs
for i in `seq 1 $N`; do
  echo "foo$i();" >> lots-of-used.rs
done
echo '}' >> lots-of-used.rs
```

we find:

```
$ rustc +nightly -V
rustc 1.21.0-nightly (aac223f4f 2017-07-30)
$ sh foo.sh
$ rustc +nightly lots-of-used.rs -Z time-passes 2>&1 | grep 'name resolution'
time: 0.033; rss: 164MB name resolution
$ rustc +nightly lots-of-unused.rs -Z time-passes 2>&1 | grep 'name resolution'
time: 1.139; rss: 136MB name resolution
```

That's a lot of time spent processing unused imports!

The slowdown here isn't quite the same as the style crate, unfortunately. The style crate only has ~7k unused imports, which apparently takes it 48s in that profile. In any case the intent here is clear, lots of unused imports seem to slow down compilation, and the `style `crate looks to have a lot of unused imports!

### LINKED ISSUE #43573: Unnaturally slow expansion on generated source
In investigating https://github.com/rust-lang/rust/issues/43572 I was generating some source code, and using the shell script listed in https://github.com/rust-lang/rust/issues/43572 we can see:

```
$ rustc +nightly -V
rustc 1.21.0-nightly (aac223f4f 2017-07-30)
$ sh foo.sh
$ rustc +nightly lots-of-unused.rs -Z time-passes 2>&1 | grep 'expansion'
time: 1.567; rss: 110MB expansion
```

That's a lot of time to expand a crate that doesn't have any macros in it!

### PR DESCRIPTION
**Pull request by arielb1**
This fixes 2 problems that caused quadratic performance when lots of use-statements were present. After this patch, performance is linear (and very fast) even with 1M uses.

Fixes #43572.
Fixes #43573.

r? @eddyb

### CODE REVIEW

File: src/libsyntax_pos/lib.rs [hunk 1]
```diff
@@ -618,8 +618,11 @@ impl FileMap {
     /// If the hash of the input doesn't match or no input is supplied via None,
     /// it is interpreted as an error and the corresponding enum variant is set.
     /// The return value signifies whether some kind of source is present.
-    pub fn add_external_src(&self, src: Option<String>) -> bool {
+    pub fn add_external_src<F>(&self, get_src: F) -> bool
+        where F: FnOnce() -> Option<String>
+    {
         if *self.external_src.borrow() == ExternalSource::AbsentOk {
+            let src = get_src();
```
Comment (eddyb) [inline src/libsyntax_pos/lib.rs]: Oh wow, my bad, I didn't notice that it was loading the file *each time* in the original PR.

**Comment by alexcrichton** [comment]
Thanks for the quick fix @arielb1!

**Comment by froydnj** [comment]
Could this be uplifted to beta @arielb1 ?  Firefox would love having this in the 1.20 release!

**Comment by eddyb** [comment]
I agree with backporting to beta. cc @nikomatsakis

**Comment by alexcrichton** [comment]
Tagging as accepted for beta

**Comment by alexcrichton** [comment]
@arielb1 the second commit here, 70478ca5c83513beb91cce78ae57ade70849fca4, doesn't cleanly apply against beta, mind sending a PR to beta for this?

**Comment by arielb1** [comment]
> @arielb1 the second commit here, 70478ca, doesn't cleanly apply against beta, mind sending a PR to beta for this?

The second commit fixes a less important edge case. If it doesn't apply cleanly, I'll prefer not to include it.

**Pull request merged**